File: web-client/src/blocs/TodoBloc.js
import { ReplaySubject, from, concat } from 'rxjs';
import {
  tap,
  map,
  flatMap,
  scan,
  startWith,
  shareReplay,
} from 'rxjs/operators';

class Filter {
  static all = 'all';
  static active = 'active';
  static completed = 'completed';
}

class ItemOp {
  static add = 'add';
  static remove = 'remove';
  static filter = 'filter';
  static updateName = 'updateName';
  static updateCompleted = 'updateCompleted';
}

export default class TodoBloc {
  constructor(todoService) {
    this._itemSubject = new ReplaySubject();

    const allItems = from(todoService.getAll()).pipe(
      tap(_ => console.log('BOOM', _)),
      flatMap(data => data.items),
      shareReplay()
    );

    const data = {
      items: [],
      filter: Filter.all,
    };

    this._itemObservable = concat(allItems, this._itemSubject).pipe(
      flatMap(async request => {
        if (!request.op) {
          data.items.push(request);
          return data;
        }

        switch (request.op) {
          case ItemOp.add: {
            await this._addItem(todoService, data.items, request.item);
            return data;
          }

          case ItemOp.remove: {
            await this._removeItem(todoService, data.items, request.id);
            return data;
          }

          case ItemOp.updateName: {
            await this._updateName(
              todoService,
              data.items,
              request.id,
              request.name
            );
            return data;
          }

          case ItemOp.updateCompleted: {
            await this._updateCompleted(
              todoService,
              data.items,
              request.id,
              request.completed
            );
            return data;
          }

          case ItemOp.filter: {
            data.filter = request.filter;
            return data;
          }

          default:
            throw new Error(`Unknown operation: "${request.op}"`);
        }
      }),
      flatMap(async results => {
        const r = await results;
        return {
          items: this._applyFilter(r.items, r.filter),
          filter: r.filter,
          activeCount: this._applyFilter(r.items, Filter.active).length,
        };
      }),
      shareReplay()
    );

    // this._itemObservable = concat(allItems, this._itemSubject).pipe(
    //   startWith({
    //     items: [],
    //     filter: Filter.all,
    //   }),
    //   scan(async (acc, request) => {
    //     if (!request.op) {
    //       const a = await acc;
    //       a.items.push(request);
    //       return acc;
    //     }

    //     switch (request.op) {
    //       case ItemOp.add: {
    //         const a = await acc;
    //         await this._addItem(todoService, a.items, request.item);
    //         return acc;
    //       }

    //       case ItemOp.remove: {
    //         const a = await acc;
    //         await this._removeItem(todoService, a.items, request.id);
    //         return acc;
    //       }

    //       case ItemOp.updateName: {
    //         const a = await acc;
    //         await this._updateName(
    //           todoService,
    //           a.items,
    //           request.id,
    //           request.name
    //         );
    //         return acc;
    //       }

    //       case ItemOp.updateCompleted: {
    //         const a = await acc;
    //         await this._updateCompleted(
    //           todoService,
    //           a.items,
    //           request.id,
    //           request.completed
    //         );
    //         return acc;
    //       }

    //       case ItemOp.filter: {
    //         const a = await acc;
    //         a.filter = request.filter;
    //         return acc;
    //       }

    //       default:
    //         throw new Error(`Unknown operation: "${request.op}"`);
    //     }
    //   }),
    //   flatMap(async results => {
    //     const r = await results;
    //     return {
    //       items: this._applyFilter(r.items, r.filter),
    //       filter: r.filter,
    //       activeCount: this._applyFilter(r.items, Filter.active).length,
    //     };
    //   }),
    //   shareReplay()
    // );
  }

  get items() {
    return this._itemObservable.pipe(map(results => results.items));
  }

  get activeCount() {
    return this._itemObservable.pipe(map(results => results.activeCount));
  }

  get selectedFilter() {
    return this._itemObservable.pipe(map(results => results.filter));
  }

  filter(value) {
    this._itemSubject.next({
      op: ItemOp.filter,
      filter: value,
    });
  }

  add(item) {
    this._itemSubject.next({
      op: ItemOp.add,
      item,
    });
  }

  remove(id) {
    this._itemSubject.next({
      op: ItemOp.remove,
      id,
    });
  }

  updateName(id, name) {
    this._itemSubject.next({
      op: ItemOp.updateName,
      id,
      name,
    });
  }

  updateCompleted(id, completed) {
    this._itemSubject.next({
      op: ItemOp.updateCompleted,
      id,
      completed,
    });
  }

  dispose() {
    this._itemSubject.complete();
  }

  async _addItem(service, items, newItem) {
    const newItemWithId = await service.add(newItem);
    items.push(newItemWithId);
  }

  async _removeItem(service, items, id) {
    const matchedIndex = items.findIndex(item => item.id === id);

    if (matchedIndex === -1) {
      throw new Error(`Cannot find item with ID = "${id}".`);
    }

    try {
      await service.remove(id);
    } catch (error) {
      throw new Error(error);
    }

    items.splice(matchedIndex, 1);
  }

  async _updateName(service, items, id, name) {
    const matchedIndex = items.findIndex(item => item.id === id);

    if (matchedIndex === -1) {
      throw new Error(`Cannot find item with ID = "${id}".`);
    }

    const matchedItem = items[matchedIndex];
    matchedItem.name = name;

    try {
      await service.update(id, matchedItem);
    } catch (error) {
      throw new Error(error);
    }

    items[matchedIndex].name = name;
  }

  async _updateCompleted(service, items, id, completed) {
    const matchedIndex = items.findIndex(item => item.id === id);

    if (matchedIndex === -1) {
      throw new Error(`Cannot find item with ID = "${id}".`);
    }

    try {
      await service.updateCompleted(id, completed);
    } catch (error) {
      throw new Error(error);
    }

    items[matchedIndex].completed = completed;
  }

  _applyFilter(items, filter) {
    switch (filter) {
      case Filter.active:
        return items.filter(i => !i.completed);

      case Filter.completed:
        return items.filter(i => i.completed);

      default:
        // Filter.all
        return items;
    }
  }
}

File: mobile_client/lib/blocs/todo_bloc.dart
import 'dart:async';
import 'package:async/async.dart';

import '../models/todo_items.dart';
import '../services/todo_service.dart';

enum Filter {
  all,
  active,
  completed,
}

enum ItemOp {
  add,
  remove,
  updateFilter,
  updateName,
  updateCompleted,
}

class Op {
  ItemOp op;

  TodoItem item;
  String id;
  String name;
  bool completed;
  Filter filter;

  Op();

  factory Op.add(TodoItem item) => Op()
    ..op = ItemOp.add
    ..item = item;

  factory Op.remove(String id) => Op()
    ..op = ItemOp.remove
    ..id = id;

  factory Op.updateName(String id, String name) => Op()
    ..op = ItemOp.updateName
    ..id = id
    ..name = name;

  factory Op.updateCompleted(String id, bool completed) => Op()
    ..op = ItemOp.updateCompleted
    ..id = id
    ..completed = completed;

  factory Op.updateFilter(Filter filter) => Op()
    ..op = ItemOp.updateFilter
    ..filter = filter;
}

abstract class TodoBloc {
  Stream<Iterable<TodoItem>> items;
  Stream<int> activeCount;
  Stream<Filter> selectedFilter;

  EventSink<Op> _opSink;

  void dispose() => _opSink.close();

  void add(TodoItem item) {
    _opSink.add(Op.add(item));
  }

  void remove(String id) {
    _opSink.add(Op.remove(id));
  }

  void updateName(String id, String name) {
    _opSink.add(Op.updateName(id, name));
  }

  void updateCompleted(String id, bool completed) {
    _opSink.add(Op.updateCompleted(id, completed));
  }

  void updateFilter(Filter filter) {
    _opSink.add(Op.updateFilter(filter));
  }
}

class TodoBlocImpl extends TodoBloc {
  final TodoService _service;
  final StreamController<int> _activeCountController;
  final StreamController<Filter> _selectedFilterController;
  final StreamController<Op> _opController;

  List<TodoItem> _items;
  Filter _selectedFilter;

  Stream<Iterable<TodoItem>> _itemsStream;

  EventSink<Op> get _opSink => _opController;

  TodoBlocImpl(this._service)
      : _activeCountController = StreamController.broadcast(),
        _selectedFilterController = StreamController.broadcast(),
        _opController = StreamController() {
    //
    final serverItemsStream = Stream.fromFuture(_service.getAll())
        .map((data) => data.items)
        .map((items) => items..forEach((item) => _items.add(item)));

    final opsStream = _opController.stream.asyncMap(
      (op) async => await _executeOp(op),
    );

    // _itemsStream = StreamGroup.merge([
    //   serverItemsStream,
    //   opsStream,
    // ]).asBroadcastStream();
    _itemsStream = LazyStream(() {
      return StreamGroup.merge([
        serverItemsStream,
        opsStream,
      ]).asBroadcastStream();
    });

    _items = List();
    _selectedFilter = Filter.all;
  }

  void dispose() {
    _activeCountController.close();
    _selectedFilterController.close();
    super.dispose();
  }

  Stream<Iterable<TodoItem>> get items {
    return _itemsStream;
  }

  Stream<int> get activeCount {
    return _activeCountController.stream;
  }

  Stream<Filter> get selectedFilter {
    return _selectedFilterController.stream;
  }

  Future<Iterable<TodoItem>> _executeOp(Op op) async {
    print(op.op);

    switch (op.op) {
      case ItemOp.add:
        await _addItem(_service, _items, op.item);
        _updateActiveCount();
        break;

      case ItemOp.remove:
        await _removeItem(_service, _items, op.id);
        _updateActiveCount();
        break;

      case ItemOp.updateName:
        await _updateName(_service, _items, op.id, op.name);
        break;

      case ItemOp.updateCompleted:
        await _updateCompleted(_service, _items, op.id, op.completed);
        _updateActiveCount();
        break;

      case ItemOp.updateFilter:
        _selectedFilter = op.filter;
        _selectedFilterController.add(_selectedFilter);
        break;

      default:
        throw Exception('Unknown operation: ${op.op}');
    }

    return _applyFilter(_items, _selectedFilter);
  }

  void _updateActiveCount() {
    final activeCount = _applyFilter(_items, Filter.active).length;
    _activeCountController.add(activeCount);
  }

  Future<void> _addItem(
    TodoService service,
    List<TodoItem> items,
    TodoItem newItem,
  ) async {
    final newItemWithId = await service.add(newItem);
    items.add(newItemWithId);
  }

  Future<void> _removeItem(
    TodoService service,
    List<TodoItem> items,
    String id,
  ) async {
    final matchedIndex = items.indexWhere((item) => item.id == id);

    if (matchedIndex == -1) {
      throw Exception('Cannot find item with ID = "$id".');
    }

    await service.remove(id);

    items.removeAt(matchedIndex);
  }

  Future<void> _updateName(
    TodoService service,
    List<TodoItem> items,
    String id,
    String name,
  ) async {
    final matchedIndex = items.indexWhere((item) => item.id == id);

    if (matchedIndex == -1) {
      throw Exception('Cannot find item with ID = "$id".');
    }

    final matchedItem = items[matchedIndex];
    final updatedItem = TodoItem(
      id: matchedItem.id,
      name: name,
      completed: matchedItem.completed,
    );

    await service.update(id, updatedItem);

    items[matchedIndex] = updatedItem;
  }

  Future<void> _updateCompleted(
    TodoService service,
    List<TodoItem> items,
    String id,
    bool completed,
  ) async {
    final matchedIndex = items.indexWhere((item) => item.id == id);

    if (matchedIndex == -1) {
      throw Exception('Cannot find item with ID = "$id".');
    }

    await service.updateCompleted(id, completed);

    final matchedItem = items[matchedIndex];
    items[matchedIndex] = TodoItem(
      id: matchedItem.id,
      name: matchedItem.name,
      completed: completed,
    );
  }

  List<TodoItem> _applyFilter(List<TodoItem> items, Filter filter) {
    switch (filter) {
      case Filter.active:
        return items.where((i) => !i.completed).toList();

      case Filter.completed:
        return items.where((i) => i.completed).toList();

      default: // Filter.all
        return items;
    }
  }
}
